File: backend/weather_service.py
import os
import logging
import requests
from typing import Dict, Any, Optional
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)

OPENWEATHER_API_KEY = os.getenv("OPENWEATHER_API_KEY")
BASE_URL = "https://api.openweathermap.org/data/2.5"

def get_current_weather(lat: float, lon: float) -> Optional[Dict[str, Any]]:
    if not OPENWEATHER_API_KEY:
        logger.warning("OpenWeather API key not configured")
        return {
            "temp": 28.0,
            "feels_like": 30.0,
            "humidity": 65,
            "description": "Partly cloudy",
            "wind_speed": 5.5,
            "icon": "02d"
        }
    
    try:
        response = requests.get(
            f"{BASE_URL}/weather",
            params={
                "lat": lat,
                "lon": lon,
                "appid": OPENWEATHER_API_KEY,
                "units": "metric"
            },
            timeout=10
        )
        
        if response.status_code == 200:
            data = response.json()
            return {
                "temp": data["main"]["temp"],
                "feels_like": data["main"]["feels_like"],
                "humidity": data["main"]["humidity"],
                "pressure": data["main"]["pressure"],
                "description": data["weather"][0]["description"],
                "wind_speed": data["wind"]["speed"],
                "icon": data["weather"][0]["icon"]
            }
        else:
            logger.error(f"OpenWeather API error: {response.status_code}")
            return None
    except Exception as e:
        logger.error(f"Error fetching weather: {e}")
        return None

def get_forecast(lat: float, lon: float, days: int = 5) -> Optional[Dict[str, Any]]:
    if not OPENWEATHER_API_KEY:
        logger.warning("OpenWeather API key not configured")
        return {
            "forecast": [
                {
                    "date": "2024-01-01",
                    "temp_max": 32.0,
                    "temp_min": 22.0,
                    "humidity": 65,
                    "description": "Partly cloudy",
                    "rain_probability": 20
                }
            ]
        }
    
    try:
        response = requests.get(
            f"{BASE_URL}/forecast",
            params={
                "lat": lat,
                "lon": lon,
                "appid": OPENWEATHER_API_KEY,
                "units": "metric",
                "cnt": days * 8
            },
            timeout=10
        )
        
        if response.status_code == 200:
            data = response.json()
            
            daily_forecasts = {}
            for item in data["list"]:
                date = item["dt_txt"].split(" ")[0]
                
                if date not in daily_forecasts:
                    daily_forecasts[date] = {
                        "temps": [],
                        "humidity": [],
                        "description": item["weather"][0]["description"],
                        "rain_prob": item.get("pop", 0) * 100
                    }
                
                daily_forecasts[date]["temps"].append(item["main"]["temp"])
                daily_forecasts[date]["humidity"].append(item["main"]["humidity"])
            
            forecast = []
            for date, info in daily_forecasts.items():
                forecast.append({
                    "date": date,
                    "temp_max": max(info["temps"]),
                    "temp_min": min(info["temps"]),
                    "humidity": sum(info["humidity"]) / len(info["humidity"]),
                    "description": info["description"],
                    "rain_probability": info["rain_prob"]
                })
            
            return {"forecast": forecast[:days]}
        else:
            logger.error(f"OpenWeather API error: {response.status_code}")
            return None
    except Exception as e:
        logger.error(f"Error fetching forecast: {e}")
        return None

def check_weather_alerts(lat: float, lon: float) -> list:
    current = get_current_weather(lat, lon)
    forecast = get_forecast(lat, lon, days=3)
    
    alerts = []
    
    if not current or not forecast:
        return alerts
    
    if current["temp"] > 40:
        alerts.append({
            "type": "heatwave",
            "severity": "high",
            "message": f"Extreme heat warning: {current['temp']}°C. Ensure adequate irrigation."
        })
    elif current["temp"] < 5:
        alerts.append({
            "type": "frost",
            "severity": "high",
            "message": f"Frost warning: {current['temp']}°C. Protect sensitive crops."
        })
    
    for day in forecast["forecast"]:
        if day["rain_probability"] > 70:
            alerts.append({
                "type": "heavy_rain",
                "severity": "medium",
                "message": f"Heavy rain expected on {day['date']} ({day['rain_probability']}% chance). Plan accordingly."
            })
    
    if current["wind_speed"] > 15:
        alerts.append({
            "type": "strong_wind",
            "severity": "medium",
            "message": f"Strong winds: {current['wind_speed']} m/s. Secure equipment."
        })
    
    return alerts

File: src/services/weatherService.ts
const OPENWEATHER_API_KEY = import.meta.env.VITE_OPENWEATHER_API_KEY;
const BASE_URL = 'https://api.openweathermap.org/data/2.5';

export interface WeatherData {
  temp: number;
  feels_like: number;
  humidity: number;
  wind_speed: number;
  description: string;
  icon: string;
  rain_probability: number;
}

export interface ForecastData {
  date: string;
  temp_max: number;
  temp_min: number;
  description: string;
  icon: string;
  rain_probability: number;
}

export const getCurrentWeather = async (lat: number, lon: number): Promise<WeatherData> => {
  try {
    const response = await fetch(
      `${BASE_URL}/weather?lat=${lat}&lon=${lon}&appid=${OPENWEATHER_API_KEY}&units=metric`
    );
    
    if (!response.ok) {
      throw new Error('Failed to fetch weather data');
    }
    
    const data = await response.json();
    
    return {
      temp: Math.round(data.main.temp),
      feels_like: Math.round(data.main.feels_like),
      humidity: data.main.humidity,
      wind_speed: Math.round(data.wind.speed * 3.6), // Convert m/s to km/h
      description: data.weather[0].description,
      icon: data.weather[0].icon,
      rain_probability: data.clouds?.all || 0,
    };
  } catch (error) {
    console.error('Error fetching weather:', error);
    throw error;
  }
};

export const getWeatherForecast = async (lat: number, lon: number): Promise<ForecastData[]> => {
  try {
    const response = await fetch(
      `${BASE_URL}/forecast?lat=${lat}&lon=${lon}&appid=${OPENWEATHER_API_KEY}&units=metric`
    );
    
    if (!response.ok) {
      throw new Error('Failed to fetch forecast data');
    }
    
    const data = await response.json();
    
    // Group by date and get daily forecast
    const dailyForecasts: { [key: string]: any[] } = {};
    
    data.list.forEach((item: any) => {
      const date = item.dt_txt.split(' ')[0];
      if (!dailyForecasts[date]) {
        dailyForecasts[date] = [];
      }
      dailyForecasts[date].push(item);
    });
    
    // Get next 5 days
    const forecast: ForecastData[] = Object.keys(dailyForecasts)
      .slice(0, 5)
      .map(date => {
        const dayData = dailyForecasts[date];
        const temps = dayData.map(d => d.main.temp);
        const midDayData = dayData[Math.floor(dayData.length / 2)];
        
        return {
          date,
          temp_max: Math.round(Math.max(...temps)),
          temp_min: Math.round(Math.min(...temps)),
          description: midDayData.weather[0].description,
          icon: midDayData.weather[0].icon,
          rain_probability: midDayData.pop * 100, // Probability of precipitation
        };
      });
    
    return forecast;
  } catch (error) {
    console.error('Error fetching forecast:', error);
    throw error;
  }
};

export const getWeatherByCity = async (city: string): Promise<WeatherData> => {
  try {
    const response = await fetch(
      `${BASE_URL}/weather?q=${city}&appid=${OPENWEATHER_API_KEY}&units=metric`
    );
    
    if (!response.ok) {
      throw new Error('Failed to fetch weather data');
    }
    
    const data = await response.json();
    
    return {
      temp: Math.round(data.main.temp),
      feels_like: Math.round(data.main.feels_like),
      humidity: data.main.humidity,
      wind_speed: Math.round(data.wind.speed * 3.6),
      description: data.weather[0].description,
      icon: data.weather[0].icon,
      rain_probability: data.clouds?.all || 0,
    };
  } catch (error) {
    console.error('Error fetching weather:', error);
    throw error;
  }
};

export const getWeatherIcon = (iconCode: string): string => {
  return `https://openweathermap.org/img/wn/${iconCode}@2x.png`;
};
